File: scripts/backfill-blog-images.py
"""Backfill blog posts with embedded image placeholders.

For each blog text asset in Akashic:
1. Extract post_id from source record URL
2. Re-fetch blog with blog_downloader to get image positions
3. Upload images (skipped if duplicate) and collect asset IDs
4. Replace {{IMG:N}} with {{IMG:asset_id}} in body_text_with_images
5. Update the AssetText content

Usage:
    pip install blog-downloader  # or use the local package
    python scripts/backfill-blog-images.py [--dry-run] [--limit N]

Requires:
    AKASHIC_API_KEY env var (or .env file)
    AKASHIC_BASE_URL env var (default: https://akashic.sekai-nina.com)
"""

from __future__ import annotations

import argparse
import os
import re
import sys
import time
from datetime import datetime, timedelta, timezone

from dotenv import load_dotenv

load_dotenv()

from akashic import AkashicClient
from blog_downloader.api import BlogClient

AKASHIC_BASE_URL = os.environ.get("AKASHIC_BASE_URL", "https://akashic.sekai-nina.com")
AKASHIC_API_KEY = os.environ.get("AKASHIC_API_KEY", "")
BLOG_URL_PATTERN = re.compile(r"hinatazaka46\.com/s/official/diary/detail/(\d+)")
IMG_PLACEHOLDER = re.compile(r"\{\{IMG:(\d+)\}\}")
JST = timezone(timedelta(hours=9))


def extract_post_id(url: str) -> int | None:
    m = BLOG_URL_PATTERN.search(url)
    return int(m.group(1)) if m else None


def main() -> None:
    parser = argparse.ArgumentParser(description="Backfill blog images into Akashic")
    parser.add_argument("--dry-run", action="store_true", help="Don't actually update")
    parser.add_argument("--limit", type=int, default=0, help="Limit number of posts to process")
    parser.add_argument("--member", type=str, default=None, help="Filter by member name")
    args = parser.parse_args()

    if not AKASHIC_API_KEY:
        print("Error: AKASHIC_API_KEY is required")
        sys.exit(1)

    akashic = AkashicClient(AKASHIC_BASE_URL, AKASHIC_API_KEY, timeout=60)
    blog_client = BlogClient(delay=1.0)

    # Fetch all text assets with blog source records
    page = 1
    per_page = 50
    processed = 0
    updated = 0
    skipped = 0
    errors = 0

    while True:
        result = akashic.list_assets(
            kind="text",
            source_type="web",
            page=page,
            per_page=per_page,
        )
        items = result.get("items", [])
        if not items:
            break

        for item in items:
            if args.limit and processed >= args.limit:
                break

            asset_id = item.get("id")
            if not asset_id:
                continue

            # Get full asset with texts and source records
            asset = akashic.get_asset(asset_id)
            texts = asset.get("texts", [])
            source_records = asset.get("sourceRecords", [])

            # Find blog URL and extract post_id
            post_id = None
            for sr in source_records:
                url = sr.get("url", "")
                post_id = extract_post_id(url)
                if post_id:
                    break

            if not post_id:
                continue

            # Find body text
            body_text = None
            body_text_id = None
            for t in texts:
                if t.get("textType") == "body":
                    body_text = t.get("content", "")
                    body_text_id = t.get("id")
                    break

            if not body_text or not body_text_id:
                continue

            # Skip if already has resolved placeholders (non-numeric)
            if re.search(r"\{\{IMG:[a-zA-Z]", body_text):
                skipped += 1
                continue

            # Filter by member if specified
            if args.member:
                title = asset.get("title", "")
                entities = asset.get("entities", [])
                entity_names = [e.get("entity", {}).get("canonicalName", "") for e in entities]
                if args.member not in title and args.member not in " ".join(entity_names):
                    continue

            processed += 1
            title = asset.get("title", "")
            print(f"\n[{processed}] post_id={post_id} title={title[:40]}")

            # Re-fetch blog to get image positions
            try:
                post = blog_client.get_post(str(post_id), post_id, fetch_images=True)
            except Exception as e:
                print(f"  Error fetching blog: {e}")
                errors += 1
                continue

            if not post:
                print(f"  Blog post not found")
                errors += 1
                continue

            if not post.image_urls:
                print(f"  No images in blog post")
                skipped += 1
                continue

            body_with_images = post.body_text_with_images
            if not body_with_images or "{{IMG:" not in body_with_images:
                print(f"  No image placeholders in parsed text")
                skipped += 1
                continue

            # Upload images and collect asset IDs
            image_asset_ids: dict[int, str] = {}
            canonical_date = asset.get("canonicalDate")
            entities = asset.get("entities", [])
            entity_list = [
                {"entityId": e["entity"]["id"], "roleLabel": e.get("roleLabel", "")}
                for e in entities
                if e.get("entity", {}).get("id")
            ]
            source_record_list = [
                {
                    "sourceKind": sr.get("sourceKind", "url"),
                    "url": sr.get("url", ""),
                    "title": sr.get("title", ""),
                    "publisher": sr.get("publisher", ""),
                    "publishedAt": sr.get("publishedAt"),
                }
                for sr in source_records
                if sr.get("url")
            ]

            total_imgs = len(post.images)
            for i, img in enumerate(post.images):
                # Find the index in image_urls by matching the URL
                img_index = None
                for idx, url in enumerate(post.image_urls):
                    if url == img.url:
                        img_index = idx
                        break
                if img_index is None:
                    continue

                image_title = f"{title} ({i + 1}/{total_imgs})" if total_imgs > 1 else title
                print(f"  Uploading image {i + 1}/{total_imgs}: {img.filename}")

                if args.dry_run:
                    image_asset_ids[img_index] = f"dry-run-{img_index}"
                    continue

                try:
                    result = akashic.upload_bytes(
                        img.data,
                        img.filename,
                        "image/jpeg",
                        title=image_title,
                        kind="image",
                        status="organized",
                        canonical_date=canonical_date,
                        source_type="web",
                        entities=entity_list,
                        source_records=source_record_list,
                    )
                    aid = result.get("id") or result.get("existingId", "")
                    if aid:
                        image_asset_ids[img_index] = aid
                        print(f"    -> {aid}" + (" (existing)" if result.get("duplicate") else ""))
                except Exception as e:
                    print(f"    Upload error: {e}")

            if not image_asset_ids:
                print(f"  No images uploaded successfully")
                errors += 1
                continue

            # Replace {{IMG:N}} with {{IMG:asset_id}}
            def replace_placeholder(m: re.Match) -> str:
                idx = int(m.group(1))
                aid = image_asset_ids.get(idx)
                if aid:
                    return f"{{{{IMG:{aid}}}}}"
                return m.group(0)  # Keep original if no asset ID

            new_content = IMG_PLACEHOLDER.sub(replace_placeholder, body_with_images)

            print(f"  Updating text content ({len(image_asset_ids)} images resolved)")
            if not args.dry_run:
                try:
                    akashic._request(
                        "PATCH",
                        f"/texts/{body_text_id}",
                        json={"content": new_content},
                    )
                    updated += 1
                except Exception as e:
                    print(f"  Update error: {e}")
                    errors += 1
            else:
                updated += 1

        if args.limit and processed >= args.limit:
            break
        if len(items) < per_page:
            break
        page += 1

    print(f"\nDone: {processed} processed, {updated} updated, {skipped} skipped, {errors} errors")


if __name__ == "__main__":
    main()

File: src/cli/backfill-blog-images.ts
/**
 * Backfill blog posts with embedded image placeholders.
 *
 * 1. Find blog text assets in DB (sourceRecord URL contains hinatazaka46 diary)
 * 2. Call blog_downloader via Python subprocess to get image positions
 * 3. Upload images to Google Drive + generate R2 thumbnails
 * 4. Replace {{IMG:N}} with {{IMG:asset_id}} in text content
 * 5. Update asset title to new format: (author)ブログ「(title)」
 *
 * Usage:
 *   npx tsx src/cli/backfill-blog-images.ts [--dry-run] [--limit N]
 */

import { PrismaClient } from "@prisma/client";
import { google } from "googleapis";
import sharp from "sharp";
import { S3Client, PutObjectCommand } from "@aws-sdk/client-s3";
import { createHash } from "crypto";
import { execSync } from "child_process";
import "dotenv/config";

const prisma = new PrismaClient();

const R2_PUBLIC_URL = process.env.R2_PUBLIC_URL ?? "";
const BLOG_DOWNLOADER_DIR = process.env.BLOG_DOWNLOADER_DIR
  ?? `${process.env.HOME}/project/hnt42-project/blog_downloader`;

const s3 = new S3Client({
  region: "auto",
  endpoint: `https://${process.env.R2_ACCOUNT_ID}.r2.cloudflarestorage.com`,
  credentials: {
    accessKeyId: process.env.R2_ACCESS_KEY_ID ?? "",
    secretAccessKey: process.env.R2_SECRET_ACCESS_KEY ?? "",
  },
});

function getDriveClient() {
  const clientId = process.env.GOOGLE_CLIENT_ID;
  const clientSecret = process.env.GOOGLE_CLIENT_SECRET;
  const refreshToken = process.env.GOOGLE_REFRESH_TOKEN;
  if (!clientId || !clientSecret || !refreshToken) throw new Error("Google Drive not configured");
  const oauth2Client = new google.auth.OAuth2(clientId, clientSecret);
  oauth2Client.setCredentials({ refresh_token: refreshToken });
  return google.drive({ version: "v3", auth: oauth2Client });
}

async function uploadToDrive(buffer: Buffer, filename: string, mimeType: string) {
  const drive = getDriveClient();
  const folderId = process.env.GOOGLE_DRIVE_FOLDER_ID;
  const res = await drive.files.create({
    requestBody: {
      name: filename,
      ...(folderId ? { parents: [folderId] } : {}),
    },
    media: { mimeType, body: require("stream").Readable.from(buffer) },
    fields: "id,webViewLink",
    supportsAllDrives: true,
  });
  // Make publicly accessible
  await drive.permissions.create({
    fileId: res.data.id!,
    requestBody: { role: "reader", type: "anyone" },
    supportsAllDrives: true,
  });
  return { fileId: res.data.id!, webViewLink: res.data.webViewLink ?? "" };
}

async function generateR2Thumbnails(assetId: string, buffer: Buffer): Promise<string | null> {
  try {
    const gallery = await sharp(buffer)
      .resize(640, null, { withoutEnlargement: true })
      .webp({ quality: 80 })
      .toBuffer();

    const bucket = process.env.R2_BUCKET_NAME ?? "akashic-thumbnails";
    await s3.send(new PutObjectCommand({
      Bucket: bucket,
      Key: `thumbnails/${assetId}/gallery.webp`,
      Body: gallery,
      ContentType: "image/webp",
      CacheControl: "public, max-age=31536000, immutable",
    }));

    return `${R2_PUBLIC_URL}/thumbnails/${assetId}/gallery.webp`;
  } catch {
    return null;
  }
}

interface BlogData {
  post_id: number;
  title: string;
  author: string;
  body_text_with_images: string;
  image_urls: string[];
  images: Array<{ url: string; data_b64: string; filename: string }>;
}

function fetchBlogPost(postId: number): BlogData | null {
  try {
    const script = `
import json, base64, sys
sys.path.insert(0, '${BLOG_DOWNLOADER_DIR}/src')
from blog_downloader.api import BlogClient
c = BlogClient(delay=0.5)
p = c.get_post(str(${postId}), ${postId}, fetch_images=True)
if not p:
    print(json.dumps(None))
else:
    print(json.dumps({
        "post_id": p.post_id,
        "title": p.title,
        "author": p.author,
        "body_text_with_images": p.body_text_with_images,
        "image_urls": p.image_urls,
        "images": [{"url": img.url, "data_b64": base64.b64encode(img.data).decode(), "filename": img.filename} for img in p.images],
    }))
`;
    const result = execSync(`cd "${BLOG_DOWNLOADER_DIR}" && uv run python -c '${script.replace(/'/g, "'\\''")}'`, {
      timeout: 120000,
      maxBuffer: 100 * 1024 * 1024,
      encoding: "utf-8",
    });
    const parsed = JSON.parse(result.trim().split("\n").pop()!);
    return parsed;
  } catch (err) {
    console.error(`  Python error: ${(err as Error).message.slice(0, 200)}`);
    return null;
  }
}

function extractPostId(url: string): number | null {
  const m = url.match(/hinatazaka46\.com\/s\/official\/diary\/detail\/(\d+)/);
  return m ? parseInt(m[1]) : null;
}

async function main() {
  const args = process.argv.slice(2);
  const dryRun = args.includes("--dry-run");
  const limitArg = args.find(a => a.startsWith("--limit="));
  const limit = limitArg ? parseInt(limitArg.split("=")[1]) : 0;

  // Find blog text assets with hinatazaka46 diary source records
  const blogAssets = await prisma.asset.findMany({
    where: {
      kind: "text",
      sourceType: "web",
      sourceRecords: {
        some: { url: { contains: "hinatazaka46.com/s/official/diary/detail" } },
      },
    },
    include: {
      texts: { where: { textType: "body" }, take: 1 },
      sourceRecords: true,
      entities: { include: { entity: true } },
    },
    orderBy: { canonicalDate: { sort: "desc", nulls: "last" } },
  });

  console.log(`Found ${blogAssets.length} blog assets`);

  let processed = 0;
  let updated = 0;
  let skipped = 0;
  let errors = 0;

  for (const asset of blogAssets) {
    if (limit && processed >= limit) break;

    const bodyText = asset.texts[0];
    if (!bodyText) continue;

    // Extract post_id
    let postId: number | null = null;
    for (const sr of asset.sourceRecords) {
      postId = extractPostId(sr.url ?? "");
      if (postId) break;
    }
    if (!postId) continue;

    // No skip — always re-process for title/entity fixes

    processed++;
    console.log(`\n[${processed}] post_id=${postId} title=${asset.title?.slice(0, 40)}`);

    // Fetch blog from hinatazaka46.com
    const post = fetchBlogPost(postId);
    if (!post) {
      console.log("  Blog post not found or fetch failed");
      errors++;
      continue;
    }

    const newTitle = post.author && post.title
      ? `${post.author}ブログ「${post.title}」`
      : asset.title;

    if (!post.image_urls.length) {
      // No images, but still update title and text (for HTML entity fix)
      if (!dryRun) {
        await prisma.assetText.update({
          where: { id: bodyText.id },
          data: {
            content: post.body_text_with_images || bodyText.content,
            normalizedContent: (post.body_text_with_images || bodyText.content)
              .replace(/\{\{IMG:[a-zA-Z0-9_-]+\}\}/g, "")
              .toLowerCase().normalize("NFKC").replace(/[\s\u3000]+/g, " ").trim(),
          },
        });
        if (newTitle !== asset.title) {
          await prisma.asset.update({ where: { id: asset.id }, data: { title: newTitle } });
        }
      }
      console.log("  No images, updated text/title");
      updated++;
      continue;
    }

    if (!post.body_text_with_images || !post.body_text_with_images.includes("{{IMG:")) {
      console.log("  No image placeholders");
      skipped++;
      continue;
    }

    // Upload images and collect asset IDs
    const imageAssetIds: Record<number, string> = {};
    const totalImgs = post.images.length;
    const entities = asset.entities;
    const sourceRecords = asset.sourceRecords;

    for (let i = 0; i < post.images.length; i++) {
      const img = post.images[i];
      const imgIndex = post.image_urls.indexOf(img.url);
      if (imgIndex === -1) continue;

      const imageBuffer = Buffer.from(img.data_b64, "base64");
      const sha256 = createHash("sha256").update(imageBuffer).digest("hex");

      // Check duplicate
      const existing = await prisma.asset.findFirst({ where: { sha256 } });
      if (existing) {
        imageAssetIds[imgIndex] = existing.id;
        console.log(`  Image ${i + 1}/${totalImgs}: ${existing.id} (existing)`);
        // Generate R2 thumbnail if missing
        if (!existing.thumbnailUrl?.includes("r2.sekai-nina.com")) {
          const r2Url = await generateR2Thumbnails(existing.id, imageBuffer);
          if (r2Url) {
            await prisma.asset.update({ where: { id: existing.id }, data: { thumbnailUrl: r2Url } });
          }
        }
        continue;
      }

      const imageTitle = totalImgs > 1 ? `${newTitle} (${i + 1}/${totalImgs})` : newTitle!;
      console.log(`  Uploading ${i + 1}/${totalImgs}: ${img.filename}`);

      if (dryRun) {
        imageAssetIds[imgIndex] = `dry-run-${imgIndex}`;
        continue;
      }

      try {
        const driveResult = await uploadToDrive(imageBuffer, img.filename, "image/jpeg");
        const imageAsset = await prisma.asset.create({
          data: {
            kind: "image",
            title: imageTitle,
            description: "",
            status: "organized",
            sourceType: "web",
            storageProvider: "gdrive",
            storageUrl: driveResult.webViewLink,
            storageKey: driveResult.fileId,
            sha256,
            originalFilename: img.filename,
            mimeType: "image/jpeg",
            fileSize: imageBuffer.length,
            canonicalDate: asset.canonicalDate,
            entities: {
              create: entities.map(e => ({
                entityId: e.entityId,
                roleLabel: e.roleLabel ?? "",
              })),
            },
            sourceRecords: {
              create: sourceRecords.map(sr => ({
                sourceKind: sr.sourceKind,
                url: sr.url,
                title: sr.title,
                publisher: sr.publisher,
                publishedAt: sr.publishedAt,
              })),
            },
          },
        });

        // R2 thumbnail
        const r2Url = await generateR2Thumbnails(imageAsset.id, imageBuffer);
        if (r2Url) {
          await prisma.asset.update({ where: { id: imageAsset.id }, data: { thumbnailUrl: r2Url } });
        }

        imageAssetIds[imgIndex] = imageAsset.id;
        console.log(`    -> ${imageAsset.id}`);
      } catch (err) {
        console.error(`    Upload error: ${(err as Error).message.slice(0, 100)}`);
      }
    }

    // Replace {{IMG:N}} with {{IMG:asset_id}}
    let newContent = post.body_text_with_images;
    newContent = newContent.replace(/\{\{IMG:(\d+)\}\}/g, (match, n) => {
      const aid = imageAssetIds[parseInt(n)];
      return aid ? `{{IMG:${aid}}}` : match;
    });

    console.log(`  Updating text (${Object.keys(imageAssetIds).length} images resolved)`);
    if (!dryRun) {
      await prisma.assetText.update({
        where: { id: bodyText.id },
        data: {
          content: newContent,
          normalizedContent: newContent
            .replace(/\{\{IMG:[a-zA-Z0-9_-]+\}\}/g, "")
            .toLowerCase().normalize("NFKC").replace(/[\s\u3000]+/g, " ").trim(),
        },
      });
      if (newTitle !== asset.title) {
        await prisma.asset.update({ where: { id: asset.id }, data: { title: newTitle } });
        console.log(`  Updated title: ${newTitle?.slice(0, 50)}`);
      }
    }
    updated++;
  }

  console.log(`\nDone: ${processed} processed, ${updated} updated, ${skipped} skipped, ${errors} errors`);
  await prisma.$disconnect();
}

main().catch(console.error);
